sdk/attestation/azure-security-attestation/src/main/java/com/azure/security/attestation/AttestationAdministrationAsyncClient.java

```diff
@@ -0,0 +1,485 @@
+    }
+
+    /**
+     * Sets the current policy for an attestation type.
+     *
+     * @param attestationType Specifies the trusted execution environment to be used to validate the evidence.
+     * @param options Options for setPolicy API, including policy to set and signing key.
+     * @param context Context for the operation.
+     * @throws IllegalArgumentException thrown if parameters fail the validation.
+     * @throws HttpResponseException thrown if the request is rejected by server.
+     * @throws RuntimeException all other wrapped checked exceptions if the request fails to be sent.
+     * @return the response to an attestation policy operation.
+     */
+    Mono<Response<PolicyResult>> setAttestationPolicyWithResponse(AttestationType attestationType, AttestationPolicySetOptions options, Context context) {
+        // Ensure that the incoming request makes sense.
+        AttestationTokenValidationOptions validationOptions = options.getValidationOptions();
+        if (validationOptions == null) {
+            validationOptions = this.tokenValidationOptions;
+        }
+
+        final AttestationTokenValidationOptions finalOptions = validationOptions;
+
+        // Generate an attestation token for that stored attestation policy. We use the common function in
+        // PolicyResult which is used in creating the SetPolicy hash.
+        AttestationToken setToken = generatePolicySetToken(options.getAttestationPolicy(), options.getAttestationSigner());
+
+        return this.policyImpl.setWithResponseAsync(attestationType, setToken.serialize(), context)
+            .flatMap(response -> {
+                Response<AttestationTokenImpl> token = Utilities.generateResponseFromModelType(response, new AttestationTokenImpl(response.getValue().getToken()));
+                return getCachedAttestationSigners()
+                    .map(signers -> {
+                        token.getValue().validate(signers, finalOptions);
+                        PolicyResult policyResult = PolicyResultImpl.fromGenerated(token.getValue().getBody(com.azure.security.attestation.implementation.models.PolicyResult.class));
+                        return Utilities.generateAttestationResponseFromModelType(response, token.getValue(), policyResult);
+                    });
+            });
+    }
+
+    /**
+     * Calculates the PolicyTokenHash for a given policy string.
+     *
+     * The policyTokenHash is calculated by generating a policy set JSON Web Token signed by the key
+     * specified in the (optional) {@link AttestationSigningKey}.
+     *
+     * @param policy AttestationPolicy document use in the underlying JWT.
+     * @param signer Optional signing key used to sign the underlying JWT.
+     * @return A {@link BinaryData} containing the SHA-256 hash of the attestation policy token corresponding
+     * to the policy and signer.
+     */
+    public BinaryData calculatePolicyTokenHash(String policy, AttestationSigningKey signer) {
```

Comment:
>public [](http://example.com/codeflow?start=4&length=6)

Does this need to be public?

Reply:
This is the helper routine you suggested I add to allow customers to validate the hash of the policy set token. I'm not sure how customers can validate the hash of the policy token if it's not public.

Reply:
I put it on the Client object instead of the Result object because the client object already has access to the JSON serialization logic, and has its own logger. If it's a static method on the PolicyResult, then both of those need to be instantiated on every call (which is only going to happen once, but there's kinda a principle of the thing here :).

Reply:
My worry is would it be discoverable on the client classes? Should this be a Utility method instead in a helper class. But at the same time creating a new helper class for this may not be worth it. I am divided. Let's keep it here for now, we may get an opinion from architects but lets see :)